### PR TITLE
Fix TEGroupedMLP pre-backward unshard in fine-grained FSDP hooks for …

### PR DESCRIPTION
**Pull request by rapatel**
…1F1B EP overlap.

Dev PR - [Megatron-LM/pull/4989](https://github.com/NVIDIA/Megatron-LM/pull/4989)

- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by rapatel** [comment]
@cspades PTAL. Thanks!

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 1]
```diff
@@ -874,6 +887,13 @@ def _pre_backward_param_unshard(module: nn.Module, *unused):
 
             if isinstance(module, tuple(fsdp_unit_modules)):
                 param_list = list(module.parameters())
+            elif (
+                self.enable_fine_grained_param_gather_backward_hook
+                and self.fine_grained_pre_backward_recurse_module_types
+                and isinstance(module, self.fine_grained_pre_backward_recurse_module_types)
+            ):
+                # Container modules (e.g. TEGroupedMLP): sharded weights are on children.
+                param_list = list(module.parameters(recurse=True))
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: This is at least asymmetric with https://github.com/NVIDIA/Megatron-LM/blob/859b7197f19119f53113ab6d30d1901b87ca1197/megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py#L758-L759

Can you clarify the intention? I've yet to understand the combination of fine-grained **and** recurse. When enable_fine_grained_param_gather_backward_hook is on, I assume you want to register hooks on the children modules in addition to the root. Therefore, children modules are supposed to take care of their own parameters and `recurse=False`.

**Review by wujingyue — COMMENTED** [review]
Thanks for the PR!

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 1 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Good question - the confusion is mostly about what “fine-grained” registers vs what actually runs in the 1F1B overlap schedule.

Fine-grained mode registers a pre-bwd hook on every submodule so overlap can call attn / mlp / experts directly instead of TransformerLayer.forward(). For normal modules, recurse=False is right - only unshard that module’s own parameters.

TEGroupedMLP is a container with no direct parameters (weights are on linear_fc1 / linear_fc2), so recurse=False gives an empty param list and the hook does nothing. Child linears don’t save us in the fused/overlap path because their forward often doesn’t run; backward is tied to TEGroupedMLP’s output, so that hook must use recurse=True. This is an allowlist for that container case, not “recurse everywhere.”

This PR is to allow the 1F1B implementation from @Wohox - https://github.com/NVIDIA/Megatron-LM/pull/3797 - to run with fused version of TEGroupedMLP. Without this change, we miss the un-sharding (param AG) of linear fc1/2 weights during bprop.

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: > TEGroupedMLP is a container with no direct parameters (weights are on linear_fc1 / linear_fc2), so recurse=False gives an empty param list and the hook does nothing. Child linears don’t save us in the fused/overlap path because their forward often doesn’t run

Interesting -- I saw a similar pattern with MambaLayer. #4899 is for fixing that because we own the source code. 

How does this PR fix this problem? If hooks are attached to TEGroupedMLP, communication is triggered before/after TEGroupedMLP and therefore not fine-grained enough for overlapping. If to submodules, the parameters may be used outside the unshard-to-reshard window, which is unsafe. What am I missing?

Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Thanks I believe this is the similar idea as MambaLayer (#4899) (container + fine-grained hooks), but for MoE experts under overlap.

I am not adding new hooks - with enable_fine_grained_param_gather_backward_hook, hooks are already on every submodule, including TEGroupedMLP. The bug is the param list - recurse=False on the container -> no params -> no unshard -> MMU fault. We only fix that list so that we grab the appropriate params.

Overlap (1F1B) already runs at submodule granularity (experts = TEGroupedMLP). On the fused path, linear_fc1/linear_fc2 don’t run forward, so backward only sees the TEGroupedMLP hook. We’re not making overlap coarser - we’re making that hook actually gather child weights.

Also, we’re not moving hooks to children only. Child hooks may exist but don’t run if their forward didn’t run. The failure was no unshard before expert backward, not params used outside the window.

@cspades Please chime in as I may be missing something and you folks are a lot more familiar with this codebase. Want to make sure my understanding is correct. Thanks!

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Got it -- there are actually three levels: the parent of TEGroupedMLP (which is an FSDP unit), TEGroupedMLP, and the direct children of TEGroupedMLP. TEGroupedMLP is actually at the correct FSDP granularity. 

I'll review this again soon and thanks for the context!

Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Sure - thanks for reviewing it for me.

BTW - if you take a look at the [1f1b schedule logic](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/models/common/model_chunk_schedule_plan.py#L230) - you can see which module is getting invoked at what time. You'll see the following:

```python
        if b_layer is not None:
            b_grad = b_layer.mlp.backward(b_grad)
```

The fsdp unit is `transformer_layer` (parent of TEGroupedMLP), but the unit itself is never invoked.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 2]
```diff
@@ -177,6 +177,12 @@ class MegatronFSDP(torch.nn.Module):
             userbuffer registration when nccl_ub is set.
         enable_fine_grained_param_gather (bool): Whether to enable "fine-grained" param all-gather,
             which can improve performance when using MXFP8 parameters with activation recomputation.
+        enable_fine_grained_param_gather_backward_hook (bool): Register pre-backward unshard hooks
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:180]: Note to myself: this PR doesn't add this flag; it merely adds a missing comment.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 1 above)
Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: The PR makes more sense to me now. 

These are probably pre-existing issues, but what still confuse me:
1. There's an asymmetry between pre_forward_param_unshard and pre_backward_param_unshard: https://github.com/NVIDIA/Megatron-LM/blob/146d1719f8cf6631ddd325c40a9ed72710a1a986/megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py#L753-L760. Shouldn't this change apply to pre-forward as well otherwise linear_fc1/2's parameters might not be ready before forward? Is the problem being hidden by prefetching again? 
2. Why not recurse=True unconditionally? The comment [here](https://github.com/NVIDIA/Megatron-LM/blob/146d1719f8cf6631ddd325c40a9ed72710a1a986/megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py#L754-L756) is plausible. However, don't we have a correctness issue whenever `module` is a submodule of an FSDP unit and doesn't directly own every parameter under it? 
3. Looks like unshard is fine-grained but reduce_grads is not (it's still on TransformerLayer). Does that create a problem for 1f1b overlapping? 

Regarding the PR, if you could rename fine_grained_pre_backward_recurse_module_types to fine_grained_recurse_module_types and apply it both pre-forward and pre-backward, the PR will LGTM. The blast radius would be small because fine_grained_pre_backward_recurse_module_types is a new opt-in knob. 

cc @shjwudp

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 3]
```diff
@@ -553,6 +562,25 @@ def _register_fsdp_hooks(self, root_module):
         """
         fsdp_unit_modules = self.fsdp_unit_modules
 
+        def _param_list_for_submodule_unshard(
+            module: nn.Module, *, for_backward: bool
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Consider https://typing.python.org/en/latest/spec/literal.html instead of a boolean. e.g. `Literal["forward", "backward"]`. We could do this for optimizer and it's not immediately clear what `for_backward=False` means. 

Ref: https://www.teamten.com/lawrence/programming/prefer-enums-over-booleans.html

**Review by wujingyue — APPROVED** [review]

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 3 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Great suggestion - applied. Thanks!

**Review by Phlip79 — APPROVED** [review]

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 4]
```diff
@@ -546,6 +555,25 @@ def _register_fsdp_hooks(self, root_module):
...
+                if pass_direction == "backward"
+                else self.enable_fine_grained_param_gather_hook
+            )
+            if (
+                fine_grained_enabled
+                and self.fine_grained_recurse_module_types
+                and isinstance(module, self.fine_grained_recurse_module_types)
+            ):
+                return list(module.parameters(recurse=True))
+            return list(module.parameters(recurse=False))
```
Comment (cspades) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Wait, looking at this again, shouldn't the logic be:

- If not an FSDP unit, then don't recurse.
  - Always fine-grained AG, since we don't want to over-gather parameters that we may not need yet, as this module doesn't have a batch specification (or user-guaranteed FSDP-compliant schedule) for how many parameters need to be communicated. **Attach hooks to _everything_!**
- If an FSDP unit...
  - If not fine-grained, then recurse.
    - This is the vanilla case, an FSDP unit module that is not fine-grained should fetch all of its parameters.
  - If fine-grained and is not a recursing module...
    - If we are gathering forward-pass / row-wise parameters (`_pre_forward_param_unshard`), then don't recurse.
      - **_Even if the module is an FSDP unit module, we want to fine-grain AG the forward pass / row-wise parameters!_** ❗ 
      - Used to deal with activation recompute (with forward pass parameters). This is because activation recompute is at a fine-grained module level (not at the FSDP unit level), so without a fine-grained hook, module forward parameters will not be gathered and thus activation recompute will be functionally incorrect.
        - Without MXFP8, backward AG == forward AG. So recompute is no-problemo, because backward AG always recurses.
        - With MXFP8, we absolutely need to fine-grain AG the row-wise parameters.
    - If we are gathering backward-pass / column-wise parameters (`_pre_backward_param_unshard`), then recurse.
      - Vanilla case. Backward pass parameters don't need fine-grained AG, we can just gather everything since we're going to use everything anyway.
  - If fine-grained and is a recursing module, then recurse.
    - Like `TEFusedMLP` where its parameters lose their membership (and hooks) to the top-level `TEFusedMLP` module. This is your new override logic. 🍀 

The logic currently looks like:

- If FSDP unit, then recurse. (`recurse=True` is the default!)
- If not FSDP unit and fine-grained and not a recursing module, then don't recurse.
- If not FSDP unit and fine-grained but it is a recursing module, then recurse.

TL;DR FSDP unit modules can still be fine-grained AG for sub-module recompute.

**Review by cspades — CHANGES_REQUESTED** [review]

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 4 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: cc @Autumn1998 to check if this logic makes sense.

Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: cc @wujingyue @shjwudp as well. We need better APIs here for all of these cases in V2.

Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: @cspades Yeah I think you're right... but I don't think anything outside of TransformerLayer right now is an FSDP unit. And in the 1F1B case, we saw the TEGroupedMLP module gets invoked, and we never recurse it for the param list, so it returns [], so we need to recurse it to grab the param list from fc1 and fc2.

Let me think about it and come up with a proposed change. Good catch though - thanks!

Comment (Autumn1998) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: I understand that during recompute, a backward pass may trigger the forward of any individual part. Because of that, we need to attach fine-grained hooks to each part of the forward, so that backward can trigger them in a fine-grained way and avoid unnecessary communication. From my perspective, the logic looks correct. For the 1F1B scenario there may be more complex issues, since we have to manually call the forward and backward of some submodules, and @shjwudp  has been working on that recently.

Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Addressed and reviewed changes with @cspades. See 338dfbb91c2af6d9c319e5b28762e2ba662b7f3a.

**Review by cspades — APPROVED** [review]

**Review by deepakn94 — APPROVED** [review]

**Comment by cspades** [comment]
/ok to test [c0422cb](https://github.com/NVIDIA/Megatron-LM/pull/4990/commits/c0422cbdef5e5b1933af10fcd9d19d171d794e86)

**Comment by cspades** [comment]
/ok to test [00ddbc6](https://github.com/NVIDIA/Megatron-LM/pull/4990/commits/00ddbc67074c40ba037bf711c6e8ec03a976ad21)

**Comment by cspades** [comment]
/ok to test [cc78b37](https://github.com/NVIDIA/Megatron-LM/pull/4990/commits/cc78b379a420e37fcb68c35d08723c54ea7295a9)
